### ISSUE TITLE
New version reports an indirect reference error

### ISSUE THREAD

**Issue by amitjoy** [groovy/groovy-eclipse#893]
The new version of Groovy Development Tools reports the following error -

`The type groovy.transform.Generated cannot be resolved. It is indirectly referenced from required .class files`

Affected Version:
  Eclipse Groovy Development Tools	`3.4.0.v201905200334-e1903`

The class `groovy.transform.Generated` is available in 2.5.x+. If I change the Groovy Compiler version to use 2.4.x in Eclipse, the workspace crashes. It never starts again. It is reproducible too.

I uninstalled Groovy Compiler 2.5 so that Eclipse always uses the available 2.4 version but the aforementioned error message still exists.

**Comment by davidmichaelkarr** [comment]
I am seeing this error with my new 2021-03 installation.  What is the proper user-level fix for this?  The original post describes two workarounds that sound like they just create other problems.

**Comment by eric-milles** [comment]
@davidmichaelkarr Groovy 2.4 support was dropped in release 4.0.0.  If you want to continue using the Groovy 2.5 compiler on a project with Groovy 2.4 or earlier libraries, you could add a definition for groovy.transform.Generated or you could stick with release 3.9.0.  See also #1235.

**Comment by davidmichaelkarr** [comment]
I'm fine with 2.4, I just needed a path forward to resolve this. The information in the issue up to this point was ambiguous.

**Comment by eric-milles** [comment]
Since release 4.0.0, the compiler always adds `@Generated` to several `GroovyObject` methods for most Groovy classes.  To support this with a Groovy 2.4 runtime, you need to create or import the annotation definition.

**Comment by davidmichaelkarr** [comment]
Ah, the problem appears to be a little more complex than that. My project IS at groovy compiler 2.5.  However, my project has a dependency on groovy-all, and I only see version 2.4 for that in the artifact repository.  Am I going to have to back up to groovyeclipse 3.9.0?

**Comment by eric-milles** [comment]
`groovy-all` was switched to a POM-only artifact starting with Groovy 2.5.  So you can still reference `org.codehaus.groovy:groovy-all:2.5.x` but you need to indicate that it is pom not jar.  If you want to stick with `groovy-all:2.4.x` and Groovy-Eclipse 4.x, you need to supply a definition for `groovy.transform.Generated`.  Or you can stay with Groovy-Eclipse 3.9.0 which includes support for Groovy Compiler 2.4.

**Comment by davidmichaelkarr** [comment]
I changed the dependency to 2.5.13 and type pom, and I also changed two spock dependencies to the 2.5 variant, but it didn't fix the problem.  It still reports the "The type groovy.transform.Generated cannot be resolved. It is indirectly referenced from required .class files".  I see that when I examine the dependency hierarchy, groovy-all 2.5.13 pulls in groovy 2.4.15, which doesn't make much sense.

**Comment by davidmichaelkarr** [comment]
Can you remind me of the update site where I can get 3.9.0 from?

**Comment by eric-milles** [comment]
https://github.com/groovy/groovy-eclipse/wiki/3.9.0-Release-Notes#update-sites

**Comment by davidmichaelkarr** [comment]
Can you elaborate on what needs to be done to supply a definition for groovy.transform.Generated? I need to understand all of the options.

**Comment by eric-milles** [comment]
https://github.com/groovy/groovy-eclipse/issues/1235#issuecomment-805660498

**Comment by davidmichaelkarr** [comment]
Today I tried to install the version associated with 2020-12, but it wouldn't even let me do it. Once I selected that, it changed the request to the 2021-03 version.  I even tried disabling the e4.19 update site, but it STILL overrode the installation request  to the version associated with 2021-03, but since I disabled that update site, it then failed, saying it couldn't find the required version.
